File: i2f-jdk/i2f-match/src/main/java/i2f/match/IMatcher.java
package i2f.match;

/**
 * @author Ice2Faith
 * @date 2022/4/26 9:34
 * @desc 一个允许返回匹配字符串，并且可以得出匹配精确度的匹配接口
 * 在一般匹配结果为布尔值的情况下，增强了用于同一个串匹配多个表达式时，能够得出哪一个表达式更加匹配
 * 在某些情况下，用于根据匹配表达式得到表达式之间的优先级
 */
public interface IMatcher {
    double MATCH_FAILURE_VALUE = -1.0;
    double MATCH_SUCCESS_LIMIT = -0.5;

    /**
     * 通配符匹配
     * 返回值含义，为负数，表示不匹配
     * 为正数，表示（末尾匹配程度+总体匹配重叠度）的均值
     * 此时可以用于表示哪一个patten更能够更加精准来匹配字符串
     * 这个返回值在某些情况下很有用，比如多个匹配规则同时适用一个目标时，优选某个规则的情况
     *
     * @param str
     * @param patten
     * @return
     */
    double match(String str, String patten);

    default boolean isMatch(String str, String patten) {
        return matched(match(str,patten));
    }

    default boolean matched(double rate) {
        return rate > MATCH_SUCCESS_LIMIT;
    }

    default double calcMatchRate(int matchEndStrIndex, int matchEndPattenIndex, int strLen, int pattenLen, int matchEndMatchedCount) {
        return ((((matchEndStrIndex + matchEndPattenIndex) * 1.0 / (strLen + pattenLen)) * 0.5) + ((matchEndMatchedCount * 1.0 / strLen) * 0.5));
    }
}

File: i2f-jdk/i2f-swl/web/i2f-swl/src/i2f-turbo-web/i2f-core/match/IMatcher.js
/**
 * @return IMatcher
 * @abstract
 */
function IMatcher() {

}

IMatcher.MATCH_FAILURE_VALUE = function () {
    return -1.0
}
IMatcher.MATCH_SUCCESS_LIMIT = function () {
    return -0.5
}

/**
 * 通配符匹配
 * 返回值含义，为负数，表示不匹配
 * 为正数，表示（末尾匹配程度+总体匹配重叠度）的均值
 * 此时可以用于表示哪一个patten更能够更加精准来匹配字符串
 * 这个返回值在某些情况下很有用，比如多个匹配规则同时适用一个目标时，优选某个规则的情况
 *
 * @param str {String}
 * @param patten {String}
 * @return {double}
 */
IMatcher.prototype.match=function(str, patten){

}
/**
 *
 * @param str {String}
 * @param patten {String}
 * @return {boolean}
 */
IMatcher.prototype.isMatch=function(str, patten) {
    return this.matched(this.match(str,patten));
}
/**
 * @param rate {double}
 * @return {boolean}
 */
IMatcher.prototype.matched=function(rate) {
    return rate > IMatcher.MATCH_SUCCESS_LIMIT();
}

/**
 * @param matchEndStrIndex {int}
 * @param matchEndPattenIndex {int}
 * @param strLen {int}
 * @param pattenLen {int}
 * @param matchEndMatchedCount {int}
 * @return {double}
 */
IMatcher.prototype.calcMatchRate=function(matchEndStrIndex, matchEndPattenIndex, strLen, pattenLen, matchEndMatchedCount) {
    return ((((matchEndStrIndex + matchEndPattenIndex) * 1.0 / (strLen + pattenLen)) * 0.5) + ((matchEndMatchedCount * 1.0 / strLen) * 0.5));
}

export default IMatcher